web/src/main/java/org/phoenixctms/ctsms/web/component/datatable/ColumnManagementBean.java

```diff
@@ -0,0 +1,233 @@
+		getVisibleMap(dataTableClientId, null);
+	}
+
+	public static void resetVisibleMap(String dataTableId) {
+		DataTable dataTable = (DataTable) WebUtil.findComponentById(dataTableId);
+		if (dataTable != null) {
+			WebUtil.getSessionScopeBean().getColumnManager().clearVisibleMap(":" + dataTable.getClientId());
+		}
+	}
+
+	private void updateVisibleMap(String dataTableClientId, Collection<DataTableColumnVO> tableColumnVOs) {
+		getVisibleMap(dataTableClientId, tableColumnVOs);
+	}
+
+	private Map<String, String> getVisibleMap(String dataTableClientId, Collection<DataTableColumnVO> tableColumnVOs) {
+		if (tableColumnVOs == null && cache.containsKey(dataTableClientId)) {
+			return cache.get(dataTableClientId);
+		} else {
+			if (tableColumnVOs == null) {
+				try {
+					tableColumnVOs = WebUtil.getServiceLocator().getUserService().getDataTableColumns(WebUtil.getAuthentication(), WebUtil.getUser().getId(), dataTableClientId,
+							null);
+				} catch (ServiceException e) {
+				} catch (AuthenticationException e) {
+					WebUtil.publishException(e);
+				} catch (AuthorisationException e) {
+				} catch (IllegalArgumentException e) {
+				}
+			}
+			HashMap<String, Boolean> savedTableColumns;
+			LinkedHashMap<String, String> tableColumns;
+			if (tableColumnVOs != null) {
+				savedTableColumns = new HashMap<String, Boolean>(tableColumnVOs.size());
+				Iterator<DataTableColumnVO> it = tableColumnVOs.iterator();
+				while (it.hasNext()) {
+					DataTableColumnVO tableColumnVO = it.next();
+					savedTableColumns.put(tableColumnVO.getColumnName(), tableColumnVO.getVisible());
+				}
+				tableColumns = new LinkedHashMap<String, String>(savedTableColumns.size());
+			} else {
+				savedTableColumns = new HashMap<String, Boolean>();
+				tableColumns = new LinkedHashMap<String, String>();
+			}
+			//if (tableColumns.size() == 0) {
+			DataTable dataTable = (DataTable) WebUtil.findComponentByClientId(dataTableClientId);
+			if (dataTable != null) {
+				Iterator<Column> it = dataTable.getAllColumns().iterator();
+				while (it.hasNext()) {
+					Column column = it.next();
+					if (column.isRendered()) {
+						String headerTextEl = null;
+						try {
+							headerTextEl = getColumnName(column);
+						} catch (Exception e) {
+						}
+						if (headerTextEl != null) {
+							if (savedTableColumns.containsKey(headerTextEl)) {
+								tableColumns.put(headerTextEl, Boolean.toString(savedTableColumns.get(headerTextEl)));
+							} else {
+								tableColumns.put(headerTextEl, getColumnVisibleDefault(column));
+							}
+						}
+					}
+				}
+			}
+			//}
+			cache.put(dataTableClientId, tableColumns); //thread-safe enough.
+			return tableColumns;
+		}
+	}
+
+	private static String getColumnName(Column column) {
+		String headerTextEl = ((HtmlOutputText) column.getFacet("header")).getValueExpression("value").getExpressionString();
+		return headerTextEl.substring(2, headerTextEl.length() - 1); // Remove #{}
```

Comment:
not sure what's the point. any p:column strictly contains a header facet with an EL expression for the outputtext value, denoting the key in a resourcebundle.
```<p:column
			sortBy="#{ecrf.vo.name}"
			filterBy="#{ecrf.vo.name}">
			<f:facet name="header">
				<h:outputText value="#{triallabels.ecrf_list_name_column}" />
			</f:facet>
			<h:outputText value="#{ecrf.vo.name}" />
		</p:column>```
so it will always be in the format "#{bundle.labelkey}". this convention is utilized here and also to have a unique identify for a datatable column.